src/main/c/NativePtr.cpp

```diff
@@ -0,0 +1,366 @@
+  long get_addrB() const;
+  void set_addrA(long v);
+  void set_addrB(long v);
+};
+
+class NativePtrInfo {
+public:
+  static const int kMagic = 0xa35e72bf;
+  int  magic_;
+  jfieldID addrA_id_;
+  jfieldID addrB_id_;
+  
+public:
+  NativePtrInfo(JNIEnv* env, int line) {
+    auto cl = env->FindClass("is/hail/nativecode/NativeBase");
+    addrA_id_ = env->GetFieldID(cl, "addrA", "J");
+    addrB_id_ = env->GetFieldID(cl, "addrB", "J");
+    // Check that std::shared_ptr matches our assumptions
+    auto ptr = std::make_shared<NativeObj>();
+    AlignedBuf* buf = reinterpret_cast<AlignedBuf*>(&ptr);
+    assert(sizeof(ptr) == 2*sizeof(long));
+    assert(buf->get_addrA() == reinterpret_cast<long>(ptr.get()));
+    magic_ = kMagic;
+  }
+};
+
+// WARNING: I observe that when this gets loaded as a shred library on Linux,
+// we see two distinct NativePtrInfo objects at different addresses.  That is
+// extremely weird, but by putting all the initialization into the constructor
+// we make sure that both get correctly initialized.  But that behavior might
+// cause trouble in other code, so we need to watch out for it.
+
+static NativePtrInfo* get_info(JNIEnv* env, int line) {
+  static NativePtrInfo the_info(env, line);
+  return &the_info;
+}
+
+// We use non-inline methods to try to defeat over-aggressive reordering
+// of aliased type-punned accesses.  The cost is probably small relative
+// to the overhead of the Scala-to-C++ JNI call.
+
+long AlignedBuf::get_addrA() const { return *(long*)(&buf_[0]); }  
+long AlignedBuf::get_addrB() const { return *(long*)(&buf_[8]); }
+void AlignedBuf::set_addrA(long v) { *(long*)&buf_[0] = v; }
+void AlignedBuf::set_addrB(long v) { *(long*)&buf_[8] = v; }
+
+} // end anon
+
+NativeObj* get_from_NativePtr(JNIEnv* env, jobject obj) {
+  auto info = get_info(env, __LINE__);
+  long addrA = env->GetLongField(obj, info->addrA_id_);
+  return reinterpret_cast<NativeObj*>(addrA);
+}
+
+void init_NativePtr(JNIEnv* env, jobject obj, NativeObjPtr* ptr) {
+  auto info = get_info(env, __LINE__);
+  // Ignore previous values in NativePtr
+  AlignedBuf buf;
+  buf.as_NativeObjPtr() = std::move(*ptr);
+  env->SetLongField(obj, info->addrA_id_, buf.get_addrA());
+  env->SetLongField(obj, info->addrB_id_, buf.get_addrB());
+}
+
+void move_to_NativePtr(JNIEnv* env, jobject obj, NativeObjPtr* ptr) {
+  auto info = get_info(env, __LINE__);
+  long addrA = env->GetLongField(obj, info->addrA_id_);
+  if (addrA) {
+    // We need to reset() the existing NativePtr
+    AlignedBuf old;
+    old.set_addrA(addrA);
+    old.set_addrB(env->GetLongField(obj, info->addrB_id_));
+    old.as_NativeObjPtr().reset();
+  }
+  AlignedBuf buf;
+  buf.as_NativeObjPtr() = std::move(*ptr);
+  env->SetLongField(obj, info->addrA_id_, buf.get_addrA());
+  env->SetLongField(obj, info->addrB_id_, buf.get_addrB());
+}
+
+NATIVEMETHOD(void, NativeBase, nativeCopyCtor)(
+  JNIEnv* env,
+  jobject thisJ,
+  jlong b_addrA,
+  jlong b_addrB
```

Comment:
Each GetLongField takes approx 3ns.  Not as bad as SetLongField, but still my preference is
is to avoid doing the downcall-upcall-return-upcall-return ... in the cases where we don't need
it - both for performance and for simplicity of the control flow.